### ISSUE TITLE
Add "changeUser" field to entities and history to know which user last changed an entity [5]

### ISSUE THREAD

**Issue by MartinWahnschaffe** [hzi-braunschweig/SORMAS-Project#7323]
### Situation Description
Currently we only have the information on when an entity was created and changed and which user reported it. What is missing is the information who changed an entity. This can currently only be extracted from the audit database. 

In addition there is currently some redundancy between the audit database and the history tables that we want to get rid of. Our idea is to use the history tables to keep track of all changes made to entities (create, update, delete) and to provide a separate log that provides the information that a user has changed or accessed a resource without containing any actual data on the resource itself, so that we don't have any person related data in there.

### Feature Description

- [x] Add a "changeUser" field to the AbstractDomainObject and to all related database tables and history tables
- [x] update the changeUser of the AbstractDomainObject using the PrePersist and PreUpdate annotation. See User.updateJursidictionLevel

**Important**: We have to validate whether this approach actually works and only set's the user when the entity was changed. 
Otherwise we may have to detect this somehow or work with an interceptor. In this case we should have a look at the mechanics of the existing audit log.

### Additional Information

- We recently had the approach to touch all `AbstractDomainObject`s in SQL with a foreach loop: https://github.com/hzi-braunschweig/SORMAS-Project/issues/7303#issuecomment-1015185867
- For now we should not include this in the DTO classes and especially don't have to include it in the mobile app sync process. This is a backend only functionality.

**Comment by JonasCir** [comment]
I think we can use 
```sql
ALTER TABLE cases ADD COLUMN change_user BIGINT,
    ADD CONSTRAINT fk_cases_change_users
    FOREIGN KEY (change_user)
    REFERENCES users (id);
```
Together with the loop [script](https://github.com/hzi-braunschweig/SORMAS-Project/issues/7303#issuecomment-1015185867)

**Comment by JonasCir** [comment]
@MartinWahnschaffe 
I don't think `PrePersist` and `PreUpdate` will work here as `CurrentUserService::getCurrentUser` is an EJB method. 

One idea that I have would be to put this in the save method of `AbstractBaseEjb` where we have access to the current user and push the usage of `BaseFacade` across the code base.

I'm aware this is a big change but:
1. I think we need to do it anyway at some point in time to keep this project maintainable
2. We talked about pushing the `BaseFacade` before in the (related) context of generating the FHIR logs consistently
3. I think this will allow us the really streamline the code base

I would work on a POC based on the entities already managed by the new base logic and put it in the generic code there. Then we can see how to progress.

**Comment by JonasCir** [comment]
I made really good experience with introducing the base logic into the infrastructure module as well. Gave me significant reduction of code size.

**Comment by JonasCir** [comment]
Blocked on #7113

**Comment by JonasCir** [comment]
@MartinWahnschaffe We can also put it in `BaseAdoService::persist` :D

**Comment by JonasCir** [comment]
Ok I have to start from scratch, something, is horribly broken:
We can tick the first box:
```sql
do
$$
    declare
        entitytable varchar;
    begin
        for entitytable in (select c1.table_name from  information_schema."columns" c1
                                                           left join information_schema."columns" c2 on c2.table_name = c1.table_name
                                                           left join information_schema."columns" c3 on c3.table_name = c1.table_name
                            where c1.table_schema = 'public' and c1.column_name = 'uuid'
                              and c2.table_schema = 'public' and c2.column_name = 'changedate'
                              and c3.table_schema = 'public' and c3.column_name = 'creationdate'
                            order by c1.table_name asc)
            loop
                raise notice 'ALTER TABLE % ADD COLUMN change_user BIGINT,
                                ADD CONSTRAINT fk_change_user_id
                                FOREIGN KEY (change_user)
                                REFERENCES users (id);', entitytable;

                raise notice '';
            end loop;
    end;
$$;

```

**Comment by JonasCir** [comment]
We need better support for retrieving the current user attached to a session. This is blocked for now.

**Comment by vidi42** [comment]
Blocked by https://github.com/hzi-braunschweig/SORMAS-Project/issues/7813

**Comment by MartinWahnschaffe** [comment]
There is a join explosion probably and thus we reach the maximum limit of columns in postgres:
org.postgresql.util.PSQLException: ERROR: target lists can have at most 1664 entries. See #8373

**Comment by MartinWahnschaffe** [comment]
The getChangeUser method needs to be lazy.
I'm not sure why this does not happen for the automatic tests, though.

**Comment by abrudanancuta** [comment]
Ticket reopene by the test case:

1. As national user A, I already have a case created in the system(here the change user id is the NatUser A for case)
2. As national user B, I create a contact on which I connect to the case from national user A

Actual results: 
At this step the table of cases, for the case created by NatUser A will update the field of change user id with NatUser B

Expected results:
The field of change user id should remain with NatUser A

**Comment by JonasCir** [comment]
Hm i think it is up for debate what we do in this case? I mean if A creates a case and B adds a sample for, it is somewhat true that B modified the case as now it has a new sample attached to it.

I mean if I open the case and add a sample directly, it is clear that I modified it. If I do the same from the sample page and pick a case, it might not be as obvious, however, the result is the same. In both cases the case gets modified in the same way.

@MartinWahnschaffe can we discuss this tomorrow?

**Comment by MartinWahnschaffe** [comment]
The problem here is that whenever something related to a case (e.g. a related contact, sample, etc.) is changed, the onCaseChanged method is called and this will set the completeness to null so that it's updated lazily afterwards.

Ideally we would somehow exclude changes to the case completness from updates to the change user, but that's quite tricky. Let's have a call.

**Comment by MartinWahnschaffe** [comment]
@abrudanancuta Looks like that implementation just visualizes something that is happening in general and will need to be discussed and decided about in a separate issue that I need to create.
Setting the completeness to null to force a refresh is not an action that should be considered a change to the entity.

Besides that the whole on...Changes logic is quite extensive and needs to be refined to only execute the needed parts. E.g. when saving a contact we currently call the method with the unchanged case. This should not longer happen - instead we should identify the parts that are relevant for a related contact and should split them into a seperate method.

**Comment by vidi42** [comment]
@MartinWahnschaffe does the onChange logic need to be refined as part of this ticket? or can we open a separate ticket for that as well?

**Comment by vidi42** [comment]
@MartinWahnschaffe  @JonasCir Then please open the new tickets, link them here and if there is nothing else, let's close this ticket.

**Comment by MateStrysewske** [comment]
Completeness changes have been moved to #8502.
